File: all-submissions/0103-binary-tree-zigzag-level-order-traversal/solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<List<Integer>> zigzagLevelOrder(TreeNode root) {
        List<List<Integer>> answer = new ArrayList<>();
        if (root == null) return answer;
        Queue<TreeNode> queue = new LinkedList<>();
        queue.add(root);
        int currLevel = 1;

        while (!queue.isEmpty()) {
            int nodesInCurrentLevel = queue.size();
            // do some logic here for the current level
            LinkedList<Integer> eachLevelList = new LinkedList<>();
            for (int i = 0; i < nodesInCurrentLevel; i++) {
                TreeNode node = queue.remove();
                if (currLevel % 2 == 0) {
                    eachLevelList.addFirst(node.val);
                } else {
                    eachLevelList.addLast(node.val);
                }
                if (node.left != null) {
                    queue.add(node.left);
                }
                if (node.right != null) {
                    queue.add(node.right);
                }
            }
            answer.add(eachLevelList);
            currLevel++;
        }
        return answer;
        
    }
}

File: all-submissions/0103-binary-tree-zigzag-level-order-traversal/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def zigzagLevelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        isReverse = False
        if root is None:
            return []
        queue = deque()
        queue.append(root)
        result = []
        while(queue):
            node_len = len(queue)
            res = []
            if isReverse:
                for i in range(node_len-1, -1, -1):
                    res.append(queue[i].val)
            else:
                for i in range(0, node_len):
                    res.append(queue[i].val)
            result.append(res)
            for i in range(node_len):
                node = queue.popleft()
                if node.left:
                    queue.append(node.left)
                if node.right:
                    queue.append(node.right)
            isReverse = not isReverse
        return result


